Compile/src/Producers/LeptonSFProducer.cc

```diff
@@ -468,3 +470,95 @@ void LeptonTriggerSFProducer::Produce(ZJetEvent const& event,
+    }
+
+    GetEtaAxis2D(m_etarev2D,histoname);
+    m_absoluteeta   = m_etarev2D[0];
+    m_reversed_axes = m_etarev2D[1];
+    m_2D            = m_etarev2D[2];
+    if (m_reversed_axes)
+    {
+        m_etabins = &m_ybins;
+        m_ptbins = &m_xbins;
+    }
+    else
+    {
+        m_etabins = &m_xbins;
+        m_ptbins = &m_ybins;
+    }
+    // Get file
+    LOG(INFO) << "Loading lepton scale factors for Reco " << m_id << ": File " << m_sffile
+              << ", Histogram " << histoname;
+    TFile file(m_sffile.c_str(), "READONLY");
+    TH2F* sfhisto = (TH2F*)file.Get(histoname.c_str());
+
+    // Get the pT and eta bin borders
+    for (int ix = 0; ix <= sfhisto->GetNbinsX(); ++ix){
+        m_xbins.emplace_back(2 * sfhisto->GetXaxis()->GetBinCenter(ix) -
+                             sfhisto->GetXaxis()->GetBinLowEdge(ix));
+    }
+    for (int iy = 0; iy <= sfhisto->GetNbinsY(); ++iy){
+        m_ybins.emplace_back(2 * sfhisto->GetYaxis()->GetBinCenter(iy) -
+                             sfhisto->GetYaxis()->GetBinLowEdge(iy));
+    }
+    // Fill the m_sf array with the values from the root histo
+    for (int ix = 1; ix <= sfhisto->GetNbinsX(); ix++) {
+        for (int iy = 1; iy <= sfhisto->GetNbinsY(); iy++) {
+            m_sf[ix - 1][iy - 1] = static_cast<float>(sfhisto->GetBinContent(ix, iy));
+            m_er[ix - 1][iy - 1] = static_cast<float>(sfhisto->GetBinError(ix, iy));
+        }
+    }
+    file.Close();
+}
+void LeptonRecoSFProducer::Produce(ZJetEvent const& event,
+                                    ZJetProduct& product,
+                                    ZJetSettings const& settings) const
+{
+    if(product.m_zValid){
+        float sf1 = 1/GetScaleFactor(0, *product.m_zLeptons.first);
+        float sf2 = 1/GetScaleFactor(0, *product.m_zLeptons.second);
+        product.m_weights["zl1RecoSFWeight"] = sf1;
+        product.m_weights["zl2RecoSFWeight"] = sf2;
+        product.m_weights["leptonRecoSFWeight"] = sf1*sf2;
+        if (settings.GetLeptonSFVariation() == true) {
+            // up has -1 since it's inside the denominator, down +1
+            float sf1u = 1/GetScaleFactor(-1, *product.m_zLeptons.first);
+            float sf1d = 1/GetScaleFactor(+1, *product.m_zLeptons.first);
+            float sf2u = 1/GetScaleFactor(-1, *product.m_zLeptons.second);
+            float sf2d = 1/GetScaleFactor(+1, *product.m_zLeptons.second);
+            product.m_weights["zl1RecoSFWeightUp"] = sf1u;
+            product.m_weights["zl1RecoSFWeightDown"] = sf1d;
+            product.m_weights["zl2RecoSFWeightUp"] = sf2u;
+            product.m_weights["zl2RecoSFWeightDown"] = sf2d;
+            // proper error propagation needed
+            product.m_weights["leptonRecoSFWeightUp"] = sf1u*sf2u;
+            product.m_weights["leptonRecoSFWeightDown"] = sf1d*sf2d;
+        }
+    }
+    else {
+        product.m_weights["leptonRecoSFWeight"] = 0;
```

Comment:
This should be `1`, since the weights are multilplied for the nominal weight. If you don't want them in the nominal `weight` object, add them to `m_optionalWeights` map.

Reply:
The 0 is the same implementation as for the other SFs. I can add a different commit which changes this to 1, but then I would do this for all SFs at once.

Reply:
Maybe we need to clarify why this case is checked here in the first place. When this is set to a value of `0` to filter the events, where there is no valid Z-bosson, these producers are certainly the wrong place to do so. If they are meant to set a default weight value, which shouldn't change the outcome, this value must be set to `1`.

Reply:
I think this case should never even happen, since we're running the SF Producer after we recreated the Z, it's probably safe to remove or add a Fatal/Error log stating, that a valid Z is expected.

Reply:
Yes, I agree. Throwing a Fatal/Error would be appropriate to cover edge cases.